plugins/searchonline.koplugin/main.lua

```diff
@@ -0,0 +1,313 @@
+local NetworkMgr = require("ui/network/manager")
+local RadioButtonWidget = require("ui/widget/radiobuttonwidget")
+local UIManager = require("ui/uimanager")
+local url = require("socket.url")
+local util = require("util")
+local WidgetContainer = require("ui/widget/container/widgetcontainer")
+local _ = require("gettext")
+
+if not Device:canOpenLink() then
+    return { disabled = true, }
+end
+
+local SearchOnline = WidgetContainer:extend{
+    default_search_engines = {
+        "https://html.duckduckgo.com/html?q={q}",
+        "https://google.com/search?q={q}",
+        "https://en.m.wiktionary.org/wiki/{q}",
+        "https://www.vocabulary.com/dictionary/{q}",
+        "https://www.merriam-webster.com/dictionary/{q}",
+        "https://www.dictionary.com/browse/{q}"
+    },
+    user_search_engines = G_reader_settings:readSetting("searchonline_user_search_engines", {}),
+    disabled_default_engines = G_reader_settings:readSetting("searchonline_disabled_default_engines", {}),
+}
+
+function SearchOnline:init()
+    if self.ui and self.ui.menu then
+        self.ui.menu:registerToMainMenu(self)
+    end
+    self.ui.highlight:addToHighlightDialog("12_onlinesearch", function(reader_highlight_instance)
+        return self:createHighlightMenuItem(reader_highlight_instance)
+    end)
+
+    -- Combine defaults with the user search engines
+    self:rebuildSearchEngines()
+end
+
+function SearchOnline:rebuildSearchEngines()
+    -- Build the local search_engines list from defaults
+    self.search_engines = {}
+    for _, engine in ipairs(self.default_search_engines) do
+        if not self.disabled_default_engines[engine] then
+            table.insert(self.search_engines, engine)
+        end
+    end
+    -- Add user search engines
+    for _, engine in ipairs(self.user_search_engines) do
+        table.insert(self.search_engines, engine)
+    end
+end
+
+
+function SearchOnline:getHighlight(reader_highlight_instance)
+    if not reader_highlight_instance.selected_text or not reader_highlight_instance.selected_text.text then
+        return
+    end
+
+    local query = util.cleanupSelectedText(reader_highlight_instance.selected_text.text)
+    return query
+end
+
+function SearchOnline:createHighlightMenuItem(reader_highlight_instance)
+    return {
+        text = _("Search online"),
+        enabled = Device:canOpenLink(),
+        callback = function()
+            NetworkMgr:runWhenOnline(function()
+                self:chooseSearch(self:getHighlight(reader_highlight_instance))
+            end)
+        end
+    }
+end
+
+-- combines the query string with the URL string of the chosen
+-- search engine and passes it to `Device:openLink()`
+function SearchOnline:searchOnline(query, search_engine)
+    
+    local chosen_engine = nil
+    for _, engine in ipairs(self.search_engines) do
+        if engine == search_engine then
+            chosen_engine = engine
+            break
+        end
+    end

```

Comment:
Feels a bit strange to have to do this to find again the engine (feels like your datastructures could be made better, it's sometimes tedious with Lua, I often had to build an array from a table or the reverse), you would/could probably pass the object (or some index).